### ISSUE TITLE
`\zcsetup{⟨options⟩}` → `\zcsetup[⟨type⟩]{⟨options⟩}` ?

### ISSUE THREAD

**Issue by dbitouze** [gusbrs/zref-clever#33]
It would be convenient to be able to restrict to some type of references global `⟨options⟩` setup thanks to `\zcsetup`.

As an example, one could want sections (and only sections) to be referred only by their title (`ref=title`).

Hence my suggestion to extend `\zcsetup{⟨options⟩}` into `\zcsetup[⟨type⟩]{⟨options⟩}`.

**Comment by gusbrs** [comment]
Hi @dbitouze .

Well, for the options which can be set for the type only, there's already `\zcRefTypeSetup`.

**Comment by dbitouze** [comment]
Hi @gusbrs AFAICS, `ref=title` cannot be passed to `\zcRefTypeSetup`:

```latex
\documentclass{article}
\usepackage{zref-clever}
\usepackage{zref-titleref}
\usepackage{hyperref}
\zcRefTypeSetup{section}{
  ref=title
}
\begin{document}
\section{Foo}
\label{sec:foo}
\begin{figure}
A figure.
\caption{Figure 1}
\label{fig:figure-1}
\end{figure}
A reference to \zcref{sec:foo}. \zcref[S]{fig:figure-1} shows some
interesting information.
A reference can also be made to multiple labels, as in \zcref{sec:foo,
fig:figure-1}.
\end{document}
```

**Comment by gusbrs** [comment]
I never said `ref` could. I only stated if "for the options that can be set for the type only". Alas, `ref` is (currently) not one of them.

So your point is then that `ref` should be able to be set for each type? `ref=title` is not a particularly appealing use case to motivate this, we got `\nameref` for a reason. Do you have other use cases in mind?

**Comment by gusbrs** [comment]
I've put some thought in the case of a type-specific `ref` option and, while I don't deny it would be interesting, it is technically complicated.

At `\__zrefclever_zcref:nnn` we have:

https://github.com/gusbrs/zref-clever/blob/cc759c1a30cd0385e4122695ade15e70a18d6b1f/zref-clever.dtx#L5234-L5248

Now, the sorting done by `\__zrefclever_sort_labels:` depends on `\l__zrefclever_ref_property_tl`, which is the variable set by the `ref` option. But the type-specific options are set later, at `\__zrefclever_typeset_refs:` which goes through the stack of labels of (potentially) different types.

https://github.com/gusbrs/zref-clever/blob/cc759c1a30cd0385e4122695ade15e70a18d6b1f/zref-clever.dtx#L6107-L6199

In other words, for a type-specific `ref` option, the way sorting is handled by the package would have to be revamped. Which I think is much more trouble than what the extra flexibility is worth.

So, thanks for the suggestion, but `ref` remains a general option which can be set either at `\zcsetup` or at `\zcref`.

**Comment by dbitouze** [comment]
Okay (but too bad :wink:). Thanks for the consideration anyway!

**Comment by gusbrs** [comment]
That's the result of making `\zcpageref` = `\zcref[ref=page]` (it's because of `page` that the sorting depends on `ref`). It sounded a better idea when I thought about than it does now...  ;-)  To be fair, it is not a bad concept, but it seems I haven't made things general enough to support that, and it's biting us now...
